File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        
        List<Integer> list = new ArrayList<>();

        for (int num : nums) {
            if (list.size() != 0 && list.get(list.size()-1) == num) continue;
    
            int start = 0;
            int end = list.size();
            
            while (start < end) {
                int mid = start + (end - start) / 2;
                
                if (list.get(mid) < num) start = mid + 1;
                else end = mid;
            }
            if (start == list.size()) list.add(num);
            else list.set(start, num);
        }
        return list.size();
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    const output = new Array();

    for (let num of nums) {
        let start = 0;
        let end = output.length;

        while (start < end) {
            let mid = Math.floor(start + (end - start) / 2);
            if (output[mid] < num) start = mid + 1;
            else end = mid;
        }

        if (start === output.length) output.push(num);
        else output[start] = num;
    }
    console.log(output);
    return output.length;
};